clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -405,7 +402,7 @@ public boolean hasUnsent() {
                         synchronized (deque) {
                             RecordBatch first = deque.peekFirst();
                             if (first != null) {
-                                boolean backoff = first.attempts > 0 && first.lastAttemptMs + retryBackoffMs > now;
+                                boolean backoff = first.attempts() > 0 && first.waitedTimeMs(now) < retryBackoffMs;
```

Comment:
@ijuma please take a careful look here, the comparison is inverted relative to your branch. I think it's correct this way (your branch was failing tests), but just fyi :)